### PR TITLE
improve prototype transforms kernel tests

### PR DESCRIPTION
**Pull request by pmeier**
This PR brings a few improvements to the new prototype transforms kernel tests and ports some to the new framework:

- Support PIL vs tensor mask comparison.
- Port common tests for `affine_mask` and `resize_mask` and reference tests for `affine_mask` and `affine_bounding_box`to new framework. This means the three "initial" functionals, i.e. `horizontal_flip_*`, `resize_*`, and `affine_*` are now completed for all features types and thus serve as a better reference for the remaining kernels.
- Add `KernelInfo`'s for `convert_format_bounding_box` and `convert_color_space_image_tensor` since they are already in use for the sample inputs of the other kernels and thus need to be working correctly.
- Add a dtype and device consistency test, i.e. output should have the same dtype and device as the input. As discussed offline, there are a number of bounding box kernels that don't do that. I've xfailed them for now and will send a follow-up PR to fix them.
- Some minor cleanup that I will add inline comments for

### CODE REVIEW

File: test/prototype_transforms_kernel_infos.py [hunk 1]
```diff
@@ -62,7 +67,7 @@ def wrapper(image_tensor, *other_args, **kwargs):
 
 
 def sample_inputs_horizontal_flip_image_tensor():
-    for image_loader in make_image_loaders(dtypes=[torch.float32]):
+    for image_loader in make_image_loaders(sizes=[None], dtypes=[torch.float32]):
```
Comment (pmeier) [inline test/prototype_transforms_kernel_infos.py:70]: This is an effort to keep the total number of tests moderate. Since the sample inputs should only cover all branches, we don't need to test multiple image sizes here. 

There are multiple places where I constrained the number of sample inputs below. I've checked the kernels and made sure that we cover all branches, but nothing more. The extensive tests should only happen in the reference inputs.

File: test/prototype_transforms_kernel_infos.py [hunk 2]
```diff
@@ -72,21 +77,24 @@ def reference_inputs_horizontal_flip_image_tensor():
...
-    for image_loader in make_mask_loaders(dtypes=[torch.uint8]):
+    for image_loader in make_mask_loaders(sizes=[None], dtypes=[torch.uint8]):
         yield ArgsKwargs(image_loader)
 
 
 KERNEL_INFOS.extend(
     [
         KernelInfo(
             F.horizontal_flip_image_tensor,
+            kernel_name="horizontal_flip_image_tensor",
```
Comment (pmeier) [inline test/prototype_transforms_kernel_infos.py:97]: Otherwise this appears as `hflip` and will mess up the collection.

File: test/prototype_transforms_kernel_infos.py [hunk 3]
```diff
@@ -278,26 +323,38 @@ def transform(bbox):
...
+    for mask_loader, center in itertools.product(
+        make_mask_loaders(dtypes=[torch.uint8]),
+        [None, (0, 0)],
+    ):
+        yield ArgsKwargs(mask_loader, center=center, **_AFFINE_KWARGS[0])
+
+
+@pil_reference_wrapper
+def reference_affine_mask(*args, **kwargs):
+    return F.affine_image_pil(*args, interpolation=F.InterpolationMode.NEAREST, **kwargs)
```
Comment (pmeier) [inline test/prototype_transforms_kernel_infos.py:348]: Previously we had a custom implementation for the reference (see below). @vfdev-5 was there a reason to not rely on PIL as reference here?

File: test/test_prototype_transforms_functional.py [hunk 4]
```diff
@@ -653,60 +578,6 @@ def test_correctness_affine_bounding_box_on_fixed_input(device):
     torch.testing.assert_close(output_boxes.tolist(), expected_bboxes)
 
 
-@pytest.mark.parametrize("angle", [-54, 56])
-@pytest.mark.parametrize("translate", [-7, 8])
-@pytest.mark.parametrize("scale", [0.89, 1.12])
-@pytest.mark.parametrize("shear", [4])
-@pytest.mark.parametrize("center", [None, (12, 14)])
-def test_correctness_affine_mask(angle, translate, scale, shear, center):
```
Comment (pmeier) [inline test/test_prototype_transforms_functional.py:661]: See above

File: test/test_prototype_transforms_functional.py [hunk 5]
```diff
@@ -505,6 +471,41 @@ def test_eager_vs_scripted(functional_info, sample_input):
...
+                    }
+                    else []
+                )
+            ],
+        )
+        for functional_info in FUNCTIONAL_INFOS
+        for idx, sample_input in enumerate(functional_info.sample_inputs())
+    ],
+)
+def test_dtype_consistency(functional_info, sample_input):
```
Comment (pmeier) [inline test/test_prototype_transforms_functional.py:501]: This test was also added on the old framework, because we haven't ported all kernels yet, but we need to make sure the behavior is consistent for the whole API.

File: test/test_prototype_transforms_kernels.py [hunk 6]
```diff
@@ -92,7 +94,7 @@ class TestCommon:
     sample_inputs = pytest.mark.parametrize(
         ("info", "args_kwargs"),
         [
-            pytest.param(info, args_kwargs, id=f"{info.kernel.__name__}")
+            pytest.param(info, args_kwargs, id=f"{info.kernel_name}-")
```
Comment (pmeier) [inline test/test_prototype_transforms_kernels.py:97]: The extra `-` makes it more clear that the number is related to the sample input rather than to the kernel

`test_against_reference[resize_image_tensor-0]`

vs.


`test_against_reference[resize_image_tensor0]`

File: test/prototype_transforms_kernel_infos.py (hunk 1 above)
Comment (vfdev-5) [inline test/prototype_transforms_kernel_infos.py:70]: Looks strange `sizes=[None]`, what does that do ?

File: test/prototype_transforms_kernel_infos.py [hunk 7]
```diff
@@ -104,23 +112,34 @@ def sample_inputs_horizontal_flip_mask():
...
+    height, width = image_size
+    yield height, width
+    yield int(height * 0.75), int(width * 1.25)
+
+
 def sample_inputs_resize_image_tensor():
     for image_loader, interpolation in itertools.product(
         make_image_loaders(dtypes=[torch.float32]),
         [
-            F.InterpolationMode.NEAREST,
```
Comment (vfdev-5) [inline test/prototype_transforms_kernel_infos.py:111]: Why did you remove this mode ?

File: test/prototype_transforms_kernel_infos.py (hunk 3 above)
Comment (vfdev-5) [reply test/prototype_transforms_kernel_infos.py:348]: The reason is that it was a correctness test and not consistency. We do not call our library to produce reference output. You can imagine a situation if both methods have changed but ci still passing.

File: test/test_prototype_transforms_functional.py [hunk 8]
```diff
@@ -528,82 +529,6 @@ def _compute_affine_matrix(angle_, translate_, scale_, shear_, center_):
     return true_matrix
 
 
-@pytest.mark.parametrize("angle", range(-90, 90, 56))
-@pytest.mark.parametrize("translate", range(-10, 10, 8))
-@pytest.mark.parametrize("scale", [0.77, 1.0, 1.27])
-@pytest.mark.parametrize("shear", range(-15, 15, 8))
-@pytest.mark.parametrize("center", [None, (12, 14)])
-def test_correctness_affine_bounding_box(angle, translate, scale, shear, center):
```
Comment (vfdev-5) [inline test/test_prototype_transforms_functional.py:536]: Where do you put this correctness test in the new test framework ?

File: test/prototype_transforms_kernel_infos.py (hunk 1 above)
Comment (pmeier) [reply test/prototype_transforms_kernel_infos.py:70]: `sizes` is a sequence of image sizes, i.e. height and width of the images to be created. By default we use

https://github.com/pytorch/vision/blob/c0911e31d367eca9600541bd634564c73753abbf/test/prototype_common_utils.py#L165-L168

where `None` means "I don't care, just something small"

https://github.com/pytorch/vision/blob/c0911e31d367eca9600541bd634564c73753abbf/test/prototype_common_utils.py#L172-L173

Together, `sizes=[None]` means, we don't want to test the full matrix on square, landscape, and portrait images here, but only on random image_sizes.

File: test/prototype_transforms_kernel_infos.py (hunk 7 above)
Comment (pmeier) [reply test/prototype_transforms_kernel_infos.py:111]: Because the only branching is happening on `BICUBIC`

https://github.com/pytorch/vision/blob/c0911e31d367eca9600541bd634564c73753abbf/torchvision/transforms/functional_tensor.py#L466-L467

Note that we are talking about the sample inputs here. The full matrix with all interpolation modes is still there for the reference inputs. L127-L131

File: test/test_prototype_transforms_functional.py (hunk 8 above)
Comment (pmeier) [reply test/test_prototype_transforms_functional.py:536]: The sample generation is happening in `reference_affine_bounding_box` (L325) while the reference function is `reference_affine_bounding_box` (L281)

File: test/prototype_transforms_kernel_infos.py (hunk 7 above)
Comment (vfdev-5) [reply test/prototype_transforms_kernel_infos.py:111]: Sorry, I do not understand anymore how it works :)

Comment (pmeier) [reply test/prototype_transforms_kernel_infos.py:111]: As discussed offline this should be nearest and bicubic.

**Review by vfdev-5 — APPROVED** [review]
LGTM, thanks @pmeier !

File: test/prototype_transforms_kernel_infos.py (hunk 3 above)
Comment (pmeier) [reply test/prototype_transforms_kernel_infos.py:348]: As discussed offline, Victor is technically right here. Still, since our PIL kernels, e.g. `affine_image_pil` mostly wrap the PIL functionaly and only juggle the parameters a bit, it is fine to use them directly as reference for now. We can always come back later and write proper reference tests.
